static/similarity.c

```diff
@@ -0,0 +1,266 @@
+#include "similarity.h"
+
+#include "float.h"
+#include "math.h"
+#include "stdlib.h"
+#include "string.h"
+
+#define FILE_ERROR "IN, OUT or ARRAY_INPUT is NULL\n"
+#define OUT_ERROR "FILE* out is NULL\n"
+#define VECTOR_ERROR "double *vector is NULL\n"
+#define RES_ERROR "double *res is NULL\n"
+#define NULL_VECTOR_MSG "Null vector found on line %zu\n"
+#define COS_DIST_VECTOR_ERROR -3
+#define COS_DIST_VECTOR_NULL -2
+#define RESULT_DIST_FORMAT "with distance = %.3lf\n"
+#define DBL_FORMAT "%.3lf, "
+#define DBL_FORMAT_LAST "%.3lf)\n"
+#define VECTOR_SIZE_ERROR "Size of vector N = 0!\n"
+#define ERRFILE stderr
+
+#define FAIL 0
+#define SUCCESS 1
+
+char *__FILE_NAME_SIMILARITY_THREADS;
+
+int get_run(FILE *in, FILE *out, FILE *array_input, double *vector,
+            double **array, size_t M, size_t N);
+
+int found_less_dist(result_t *res, long double dist, double *array);
+
+int found_same_dist(result_t *res, double *array);
+
+int compute_cos(double *vector, double *array, size_t N, result_t *res);
+
+int get_array(FILE *array_input, double **array, size_t M, size_t N);
+
+int malloc_run_free(FILE *in, FILE *out, FILE *array_input) {
+  size_t N = 0;  // size of vector (cols)
+  size_t M = 0;  // size of array (rows)
+  fscanf(array_input, "%zu %zu\n", &M, &N);
+  int status = SUCCESS;
+  double *vector = (double *)calloc(N, sizeof(double));
+  double **array = (double **)calloc(M, sizeof(double *));
+  if (!vector || !array) {
+    status = FAIL;
+  }
+  if (status) {
+    for (size_t i = 0; i < M; ++i) {
+      array[i] = (double *)calloc(N, sizeof(double));
+    }
+    status = get_run(in, out, array_input, vector, array, M, N);
+    for (size_t i = 0; i < M; ++i) {
+      free(array[i]);
+    }
+  }
+  free(array);
+  free(vector);
+  return status;
+}
+
+int copy_file_name(char *array_file) {
+  __FILE_NAME_SIMILARITY_THREADS = (char *)malloc(strlen(array_file) + 1);
```

Comment:
"The use of two underscores (`__') in identifiers is reserved for the compiler's internal use according to the ANSI-C standard."